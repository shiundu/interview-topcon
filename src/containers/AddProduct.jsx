import { connect } from 'react-redux'
import { addProduct } from '../actions'
import AddProductForm from '../components/AddProductForm'

const mapDispatchToProps = {
    handleSubmit: addProduct
}

export default connect(null, mapDispatchToProps)(AddProductForm)