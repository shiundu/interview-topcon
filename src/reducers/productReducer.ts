import { Product } from '../models';
import { Action } from '../actions/products';
import { ActionTypes } from '../actions/types';

export interface State {
    products: Product[]
};

// Define our initialState
export const initialState: State = {
    products: []
};

export function reducer(state: State = initialState, action: Action) {
    switch (action.type) {

        case ActionTypes.INITIAL_PRODUCT_LIST: {
            const products = action.payload.products.reverse();
    
            return {
                ...state,
                products
            }
        }

        case ActionTypes.ADD_PRODUCT: {
            let product = action.payload.product;
            product.id = state.products.length  + 1;
            product.imgUrl = product.imgUrl.length > 0 ? product.imgUrl : 'https://via.placeholder.com/200'

            return {
                ...state,
                products: [product, ...state.products]
            }
        }

        case ActionTypes.UPDATE_PRODUCT: {
            const prod = action.payload.product
            return {
                ...state,
                products: state.products.map(product => {
                    if (product.id === prod.id) {
                        product.name = prod.name.length > 0 ? prod.name : prod.name;
                        product.price = prod.price.length > 0 ? prod.price : product.price;
                        product.imgUrl = prod.imgUrl.length > 0 ? product.imgUrl : 'https://via.placeholder.com/200'
                        return product
                    }
                    return product;
                })
            }
        }

        case ActionTypes.DELETE_PRODUCT: {
            const { productId } = action.payload
            return {
                ...state,
                products: state.products.filter(product => product.id !== productId)
            }
        }

        default:
            return state
    }
}