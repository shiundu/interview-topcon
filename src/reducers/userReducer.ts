import { User } from '../models';
import { Action } from '../actions/auth';
import { ActionTypes } from '../actions/types';

// Define the State interface for user reducer
export interface State {
    user: User
}

// initialState
export const initialState: State = {
    user: {
        id: 0,
        firstName: '',
        lastName: '',
        email: '',
        authenticationToken: '',
        error: false
    }
}

export function reducer(state: State = initialState, action: Action) {
    switch (action.type) {

        case ActionTypes.LOGIN_USER: {
            const user = action.payload.user
    
            return {
            ...state,
            user 
            }
        }

        case ActionTypes.LOGOUT_USER: {
            return {
                ...state,
                ...initialState
            }
        }

        default:
        return state
    }
};