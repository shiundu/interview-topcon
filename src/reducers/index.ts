import { State } from './index';
import { combineReducers } from 'redux';
import * as fromProducts from './productReducer';
import * as fromUser from './userReducer';

export interface State {
  products:  fromProducts.State,
  user: fromUser.State
}

export const initialState: State = {
  products: fromProducts.initialState,
  user: fromUser.initialState
}

export const reducer = combineReducers<State>({
  products: fromProducts.reducer,
  user: fromUser.reducer
})