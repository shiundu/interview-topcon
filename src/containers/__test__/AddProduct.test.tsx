import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import AddProduct from '../AddProduct';
import { initialState } from '.././../reducers';

const mockStore = configureMockStore();
const store = mockStore({...initialState});

describe('AddProduct container test', () => {
    it('AddProduct contaiiner renders correctly', () => {
        const tree = renderer.create(
        <Provider store={store}>
              <AddProduct />
            </Provider>)
          .toJSON();

        expect(tree).toMatchSnapshot();
    });
})
