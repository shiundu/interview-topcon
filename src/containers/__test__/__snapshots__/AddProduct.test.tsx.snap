// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct container test AddProduct contaiiner renders correctly 1`] = `
<form
  className="form-inline"
  id="productForm"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="input-group mb-2 mr-sm-2"
    >
      <input
        className="form-control"
        name="name"
        onChange={[Function]}
        placeholder="Product name"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group mb-2 mr-sm-2"
    >
      <input
        className="form-control"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        type="string"
        value=""
      />
    </div>
    <div
      className="input-group mb-2 mr-sm-2"
    >
      <input
        className="form-control"
        name="imgUrl"
        onChange={[Function]}
        placeholder="Product image url"
        type="url"
        value=""
      />
    </div>
    <div
      className="input-group mb-2 mr-sm-2"
    >
      <button
        className="btn btn-primary form-control"
        type="submit"
      >
        Add product
      </button>
    </div>
  </div>
</form>
`;
